apis/capabilities/v1beta1/activedoc_types.go

```diff
@@ -0,0 +1,232 @@
+	ID *int64 `json:"activeDocId,omitempty"`
+
+	// ProviderAccountHost contains the 3scale account's provider URL
+	// +optional
+	ProviderAccountHost string `json:"providerAccountHost,omitempty"`
+
+	// ProductResourceName references the managed 3scale product
+	// +optional
+	ProductResourceName *corev1.LocalObjectReference `json:"productResourceName,omitempty"`
+
+	// ObservedGeneration reflects the generation of the most recently observed Backend Spec.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Current state of the activedoc resource.
+	// Conditions represent the latest available observations of an object's state
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions common.Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+func (o *ActiveDocStatus) Equals(other *ActiveDocStatus, logger logr.Logger) bool {
+	if !reflect.DeepEqual(o.ID, other.ID) {
+		diff := cmp.Diff(o.ID, other.ID)
+		logger.V(1).Info("ID not equal", "difference", diff)
+		return false
+	}
+
+	if o.ProviderAccountHost != other.ProviderAccountHost {
+		diff := cmp.Diff(o.ProviderAccountHost, other.ProviderAccountHost)
+		logger.V(1).Info("ProviderAccountHost not equal", "difference", diff)
+		return false
+	}
+
+	if o.ProductResourceName != other.ProductResourceName {
+		diff := cmp.Diff(o.ProductResourceName, other.ProductResourceName)
+		logger.V(1).Info("ProductResourceName not equal", "difference", diff)
+		return false
+	}
+
+	if o.ObservedGeneration != other.ObservedGeneration {
+		diff := cmp.Diff(o.ObservedGeneration, other.ObservedGeneration)
+		logger.V(1).Info("ObservedGeneration not equal", "difference", diff)
+		return false
+	}
+
+	// Marshalling sorts by condition type
+	currentMarshaledJSON, _ := o.Conditions.MarshalJSON()
+	otherMarshaledJSON, _ := other.Conditions.MarshalJSON()
+	if string(currentMarshaledJSON) != string(otherMarshaledJSON) {
+		diff := cmp.Diff(string(currentMarshaledJSON), string(otherMarshaledJSON))
+		logger.V(1).Info("Conditions not equal", "difference", diff)
+		return false
+	}
+
+	return true
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// ActiveDoc is the Schema for the activedocs API
+type ActiveDoc struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ActiveDocSpec   `json:"spec,omitempty"`
+	Status ActiveDocStatus `json:"status,omitempty"`
+}
+
+func (a *ActiveDoc) SetDefaults(logger logr.Logger) bool {
+	updated := false
+
+	// Respect 3scale API defaults
+	if a.Spec.SystemName == nil || *a.Spec.SystemName == "" {
+		tmp := activeDocSystemNameRegexp.ReplaceAllString(a.Spec.Name, "")
+		a.Spec.SystemName = &tmp
+		updated = true
+	}
+
+	// 3scale API ignores case of the system name field
+	systemNameLowercase := strings.ToLower(*a.Spec.SystemName)
+	if *a.Spec.SystemName != systemNameLowercase {
```

Comment:
I wonder if maybe forcing this at openapi validation level would be more appropriate?